S2A record protocol refactoring to use key derivation

matthewstevenson88 opened the pull request:
This PR refactors the `s2a_crypter_create` method to derive the keys and nonces from the traffic secret. It depends on #4 and #5, but it is orthogonal to the changes made in #5.

Note: because the branch s2a-encrypt is slightly out of sync with the branch s2a-rp-create, this PR may show changes to files in the directory src/core/tsi/alts/crypt. Please ignore them. They will disappear when #4 is merged with s2a-master.

cesarghali commented on test/core/tsi/s2a/s2a_test_util.cc:
```diff
@@ -105,6 +106,36 @@ std::vector<uint8_t> chacha_poly_empty_record_bytes = {
+      crypter, /** in_half_connection **/ true,
+      /** expected_sequence **/ 0, expected_traffic_secret.size(),
+      expected_traffic_secret.data(),
+      /** expected nonce size **/ 12, expected_nonce, SSL3_RT_HEADER_LENGTH);
```
Do we have a const for `12`? It can be local to this file or function.

matthewstevenson88 replied:
In the next commit, I've deduced the expected nonce size from the test vector (rather than using `12` directly).

cesarghali commented on test/core/tsi/s2a/BUILD:
```diff
@@ -21,7 +21,9 @@ grpc_package(name = "test/core/tsi/s2a")
     name = "s2a_test_util",
     srcs = ["s2a_test_util.cc"],
-    hdrs = ["s2a_test_util.h"],
+    hdrs = [
```
Is this auto-generated? If not please revert this change to keep the PR simple, unless it was inherited from an upstream merge.

matthewstevenson88 replied:
This was auto-generated. Not sure why it makes this change sometimes and leaves it others.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -86,27 +88,135 @@ static grpc_status_code s2a_tag_size(const s2a_crypter* crypter,
+ *
+ *  On success, the method returns GRPC_STATUS_OK. Otherwise, it returns an
+ * error status code and details can be found in |error_details|. **/
+static grpc_status_code derive_tls13_secret(uint16_t ciphersuite, bool is_key,
```
Done in next commit.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -86,27 +88,135 @@ static grpc_status_code s2a_tag_size(const s2a_crypter* crypter,
+      return GRPC_STATUS_FAILED_PRECONDITION;
+  }
+
+  uint8_t label[2 + 11];
```
Done in next comment.

matthewstevenson88 commented:
Thanks for the review @cesarghali! I've implemented the changes you requested.

FYI I also removed the null byte from the test vectors that started their lives as strings.

matthewstevenson88 commented:
Thanks for the quick review @cesarghali! :)
